<font size="+3"><b>Assignment 2: Linear Models and Validation Metrics</b></font>

***
* **Full Name** = David Rodriguez Barrios
* **UCID** = 30145288
***

<font color='Blue'>In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.</font>

You can use the Table of Content on the left side of this notebook to efficiently navigate within this documents.

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|           **Part 2: Regression**           |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    0.5    |
| Step 3: Implement Machine Learning   Model |     1     |
|            Step 4: Validate Mode           |     1     |
|          Step 5: Visualize Results         |     1     |
|                  Questions                 |     2     |
|             Process Description            |     4     |
|  **Part 3:   Observations/Interpretation** |   **3**   |
|           **Part 4: Reflection**           |   **2**   |
|                  **Total**                 |   **30**  |
|                                            |           |
|                  **Bonus**                 |           |
|         **Part 5: Bonus Question**         |   **4**   |

# **Part 1: Classification (14.5 marks total)**

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|                  **Total**                 |  **14.5** |

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:


## **Step 0:** Import Libraries

In [12]:
import numpy as np
import pandas as pd

## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [13]:

from yellowbrick.datasets import load_spam
X, y = load_spam()
print(X.shape, y.shape)
print(X.dtypes)
print(y.dtypes)

#import the load_spam function from the yellowbrick.datasets module
#show the shape and data types of the X and y arrays

(4600, 57) (4600,)
word_freq_make                float64
word_freq_address             float64
word_freq_all                 float64
word_freq_3d                  float64
word_freq_our                 float64
word_freq_over                float64
word_freq_remove              float64
word_freq_internet            float64
word_freq_order               float64
word_freq_mail                float64
word_freq_receive             float64
word_freq_will                float64
word_freq_people              float64
word_freq_report              float64
word_freq_addresses           float64
word_freq_free                float64
word_freq_business            float64
word_freq_email               float64
word_freq_you                 float64
word_freq_credit              float64
word_freq_your                float64
word_freq_font                float64
word_freq_000                 float64
word_freq_money               float64
word_freq_hp                  float64
word_freq_hpl                 f

## **Step 2:** Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [14]:
# TO DO: Check if there are any missing values and fill them in if necessary
print(X.isnull().sum())
X.fillna(X.mean(), inplace=True)
print(X.isnull().sum())

print(y.isnull().sum())
y.fillna(y.mean(), inplace=True)
print(y.isnull().sum())


word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [15]:
#use the train_test_split function from skleanr to create a new feature matrix named X_small and a new target vecote named y_small that contians 5% of the data
from sklearn.model_selection import train_test_split
# X_small, X_small_test, y_small, y_small_test = train_test_split(X, y, test_size=0.95, random_state=0)
# print((X_small), (y_small))


X_small, x_small_test, y_small, y_small_test = train_test_split(X, y, train_size=0.05, random_state=0)
print((X_small), (y_small))

      word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
606             0.00               0.84           0.84           0.0   
164             0.62               0.00           0.62           0.0   
1040            0.68               0.68           0.68           0.0   
3968            0.00               0.00           1.00           0.0   
209             0.34               0.42           0.25           0.0   
...              ...                ...            ...           ...   
1033            0.27               0.00           0.27           0.0   
3264            0.49               0.00           0.00           0.0   
1653            0.00               0.00           0.19           0.0   
2607            0.00               0.00           0.00           0.0   
2732            0.00               0.20           0.20           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
606            0.00            0.00              0.84     

## **Step 3:** Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets:
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

## **Step 4:** Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

## **Step 5:** Visualize Results (4 marks for steps 3-5)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [16]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
import pandas as pd

# Assuming load_spam() is a function that loads your dataset X and target y
X, y = load_spam()

results = pd.DataFrame(columns=['Dataset', 'Data Size', 'Training Accuracy', 'Validation Accuracy'])

# Initialize the Logistic Regression model
model = LogisticRegression(max_iter=2000)

# Split the dataset into training and testing subsets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Fit the model on the entire dataset X and y
model.fit(X, y)
y_pred_train = model.predict(X)
y_pred_test = model.predict(X_test)

# Calculate and print the accuracy scores
train_accuracy = accuracy_score(y, y_pred_train)
test_accuracy = accuracy_score(y_test, y_pred_test)
print("Dataset: X and y \t Data Size:", X.shape, "\t Training Accuracy:", train_accuracy, "\t Validation Accuracy:", test_accuracy)

# Append the results to the DataFrame
results = results.append({'Dataset': 'X and y', 'Data Size': X.shape, 'Training Accuracy': train_accuracy, 'Validation Accuracy': test_accuracy}, ignore_index=True)

# Fit the model on the first two columns of X
model.fit(X.iloc[:, :2], y)
y_pred_train = model.predict(X.iloc[:, :2])
y_pred_test = model.predict(X_test.iloc[:, :2])

# Calculate and print the accuracy scores
train_accuracy = accuracy_score(y, y_pred_train)
test_accuracy = accuracy_score(y_test, y_pred_test)
print("Dataset: X first two columns and y\t Data Size:", X.shape, "\t Training Accuracy:", train_accuracy, "\t Validation Accuracy:", test_accuracy)

# Append the results to the DataFrame
results = results.append({'Dataset': 'X first two columns and y', 'Data Size': X.iloc[:,:2].shape, 'Training Accuracy': train_accuracy, 'Validation Accuracy': test_accuracy}, ignore_index=True)

# Fit the model on the smaller subset X_small and y_small
X_small, X_small_test, y_small, y_small_test = train_test_split(X, y, test_size=0.95, random_state=0)
model.fit(X_small, y_small)
y_pred_train = model.predict(X_small)
y_pred_test = model.predict(X_small_test)

# Calculate and print the accuracy scores
train_accuracy = accuracy_score(y_small, y_pred_train)
test_accuracy = accuracy_score(y_small_test, y_pred_test)
print("Dataset: X_small and y_small \t Data Size:", X_small.shape, "\t Training Accuracy:", train_accuracy, "\t Validation Accuracy:", test_accuracy)

# Append the results to the DataFrame
results = results.append({'Dataset': 'X_small and y_small', 'Data Size': X_small.shape, 'Training Accuracy': train_accuracy, 'Validation Accuracy': test_accuracy}, ignore_index=True)

print(results)

Dataset: X and y 	 Data Size: (4600, 57) 	 Training Accuracy: 0.9319565217391305 	 Validation Accuracy: 0.9423913043478261
Dataset: X first two columns and y	 Data Size: (4600, 57) 	 Training Accuracy: 0.616304347826087 	 Validation Accuracy: 0.6010869565217392


C:\Users\de_ro\AppData\Local\Temp\ipykernel_3480\3267550978.py:28: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results = results.append({'Dataset': 'X and y', 'Data Size': X.shape, 'Training Accuracy': train_accuracy, 'Validation Accuracy': test_accuracy}, ignore_index=True)
C:\Users\de_ro\AppData\Local\Temp\ipykernel_3480\3267550978.py:41: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results = results.append({'Dataset': 'X first two columns and y', 'Data Size': X.iloc[:,:2].shape, 'Training Accuracy': train_accuracy, 'Validation Accuracy': test_accuracy}, ignore_index=True)


Dataset: X_small and y_small 	 Data Size: (230, 57) 	 Training Accuracy: 0.9478260869565217 	 Validation Accuracy: 0.9050343249427918
                     Dataset   Data Size  Training Accuracy  \
0                    X and y  (4600, 57)           0.931957   
1  X first two columns and y   (4600, 2)           0.616304   
2        X_small and y_small   (230, 57)           0.947826   

   Validation Accuracy  
0             0.942391  
1             0.601087  
2             0.905034  


C:\Users\de_ro\AppData\Local\Temp\ipykernel_3480\3267550978.py:55: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results = results.append({'Dataset': 'X_small and y_small', 'Data Size': X_small.shape, 'Training Accuracy': train_accuracy, 'Validation Accuracy': test_accuracy}, ignore_index=True)


## **Questions (4 marks)**
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

<font color='Green'><b>YOUR ANSWERS HERE</b></font>

# How do the training and validation accuracy change depending on the amount of data used? Explain with values.

The training and validation accurancy change depending on the amount of data used. In theory, larger data size should enable higher validation accuracy since larger data samples should enable the model to have more information to be able to build a robust model. However, the amount of feature information per data sample is also very important. This is crucial to ensure that we incorporate models that have the most possible feature classifications possible to create the model. This is demonstrated in the values obtained in the previous codeblock, for the datasets X and y and then X first two columns and y. They both have 4600 sample points, which in comparison to the X_small and y_small dataset is large since it only had 230 samples. However, X and y had 4600 samples with 57 features, while X first two columns had 4600 samples with only 2 features. This reduced data size difference is reflected in the training and validation accuracy. X and y dataset's training accuracy was 0.931957 and its validation accuracy was 0.942391. In contrast, X first two columns and y's respective training accuracy and validation accuracy is 0.616304 and 0.601087. Thus, large data samples does not necessarily mean the training and validation accuracy will do well since feature importance is arguably significantly more important at determining the success of a model's ability. Additionally, the third data set X_small and y_small only had 230 data samples with 57 features and had the highest training accuracy and a very close validation accuracy. These values were respectively, 0.947826 and 0.905034. That being said, if we compare the first and third data set since they included the same number of features thus being a better control group for this comparison, X_small and y_small had a better training accuracy than X and y. This could be because there are fewer samples that the model needs to incorporate. Thus leading to a higher accuracy on the training dataset without necessarily having to overfit the graph. In contrast though, they had opposite validation and training accuracy success. X and y's validation accuracy was higher than X_small and y_small's validation accuracy. This suggests that the model has more success on the validating data set which the model did not use to train. This could be due to the model's larger dataset leading to better regressions since the model had more information to classify with. 

# In this case, what do a false positive and a false negative represent? Which one is worse?
In this case, the load_spam() dataset has 4600 data samples with 57 features which are described when an email is spam or not given the other features. The model we constructed predicts which mail is spam and which isn't. Thus a positive is if an email is spam and a negative is if an email is not spam. A false positive would be an email which is categorized as spam however, isn't spam. A false negative is an email which is not categorized as spam, and is actually spam. A false negative would simply be able to be deleted by the user. A false positive is worse since if an email is mistakingly categorized as spam it would be blocked and thus the user may miss the important email. 






## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>DESCRIBE YOUR PROCESS HERE</b></font>

Where did you source your code?
I sourced my code from the lab's we covered in class as well as the Sci-kit Learn Open-Source website to ensure proper usage of the library. Specifically, I had to look back at how to use sklearn's functions and read over the notes provided in class.

In what order did you complete the steps?
I completed the steps in the following order: 0,1,2,3,4,5. I did it in this order to better understand that we are working with. Additionally, it makes sense to import the libraries, then view the headers of the data to understand it, then to fill empty data samples and handle them, followed by creating the model, and then performing the training and validation accuracy. This is done by calculating the model's predicted y values on the training and validation dataset and then analyzing the differences between the predicted and actual. Finally displaying the answers in a logical form. 

If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
I used generative AI to better understand the Sci-kit learn's train-test-split's difference between the hyperparameters, train_size and test_size. I used it to better understand the difference. This did not give me code therefore, I did not have to modify anything.

Did you have any challenges? If yes, what were they? If not, what helped you to be successful?
I had challenges understanding the purpose of the code we created in the lab however, this assignment helped me understand what the commands in the sklearn library are actually doing. It helped a lot to read through the documentation on Sci-kit learn as well as reading the GitHub repository for Sci-kit Learn.

# **Part 2: Regression (10.5 marks total)**

| **Question**                               | **Point** |
|--------------------------------------------|-----------|
| **Part 2: Regression**                     |           |
| Step 1: Data Input                         | 1         |
| Step 2: Data Processing                    | 0.5       |
| Step 3: Implement Machine Learning   Model | 1         |
| Step 4: Validate Mode                      | 1         |
| Step 5: Visualize Results                  | 1         |
| Questions                                  | 2         |
| Process Description                        | 4         |
| **Total**                                  | **10.5**  |

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [17]:
# TO DO: Import spam dataset from yellowbrick library
# TO DO: Print size and type of X and y

from yellowbrick.datasets import load_concrete

X, y = load_concrete()

print(X.shape, y.shape)
print(X.dtypes)
print(y.dtypes)

(1030, 8) (1030,)
cement    float64
slag      float64
ash       float64
water     float64
splast    float64
coarse    float64
fine      float64
age         int64
dtype: object
float64


## **Step 2:** Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [18]:
# TO DO: Check if there are any missing values and fill them in if necessary
print(X.isnull().sum())
X.fillna(X.mean(), inplace=True)
print(X.isnull().sum())

print(y.isnull().sum())
y.fillna(y.mean(), inplace=True)
print(y.isnull().sum())



cement    0
slag      0
ash       0
water     0
splast    0
coarse    0
fine      0
age       0
dtype: int64
cement    0
slag      0
ash       0
water     0
splast    0
coarse    0
fine      0
age       0
dtype: int64
0
0


## **Step 3:** Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Implement the machine learning model with `X` and `y`

In [19]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LinearRegression


model = LinearRegression()

# split the data into a training and testing set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

model.fit(X_train, y_train)






LinearRegression()

## **Step 4:** Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [20]:
# TO DO: ADD YOUR CODE HERE

y_pred_train = model.predict(X_train)
y_pred_test = model.predict(X_test)




#calculate the training and validation accuracy using mean squared error and R2 score

from sklearn.metrics import mean_squared_error, r2_score


train_accuracy_mse = mean_squared_error(y_train, y_pred_train)
test_accuracy_mse = mean_squared_error(y_test, y_pred_test)

train_accuracy_r2 = r2_score(y_train, y_pred_train)
test_accuracy_r2 = r2_score(y_test, y_pred_test)



print("Dataset: X and y \t Data Size:", X.shape, "\t Training MSE:", train_accuracy_mse, "\t Validation MSE:", test_accuracy_mse)
print("Dataset: X and y \t Data Size:", X.shape, "\t Training Accuracy R2:", train_accuracy_r2, "\t Validation Accuracy R2:", test_accuracy_r2)








Dataset: X and y 	 Data Size: (1030, 8) 	 Training MSE: 110.34550122934108 	 Validation MSE: 95.63533482690428
Dataset: X and y 	 Data Size: (1030, 8) 	 Training Accuracy R2: 0.6090710418548884 	 Validation Accuracy R2: 0.6368981103411242


## **Step 5:** Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [21]:
# TO DO: ADD YOUR CODE HERE


results = pd.DataFrame(columns=['Training accuracy', 'Validation accuracy'], index=['MSE', 'R2 score'])

results.loc['MSE', 'Training accuracy'] = train_accuracy_mse
results.loc['R2 score', 'Training accuracy'] = train_accuracy_r2
results.loc['MSE', 'Validation accuracy'] = test_accuracy_mse
results.loc['R2 score', 'Validation accuracy'] = test_accuracy_r2

print(results)

         Training accuracy Validation accuracy
MSE             110.345501           95.635335
R2 score          0.609071            0.636898


## **Questions (2 marks)**
1. Did using a linear model produce good results for this dataset? Why or why not?
No, the linear model did not produce good results for this dataset. The R2 score was 0.609071 for training accuracy and then 0.636898 for the validation accuracy. Ideally the R^2 score should be 1 which shows its perfectly linear. A 0 shows that its not matching the model at all. This could be attributed due to the sample datas noise which makes it appear not perfectly linear. Additionally, it could also be due to our model not being fit enough for the training data. Overall this is not ideal for a model. The MSE can only be taken into account when a general idea of the sample is. For example, a 110.345501 mean sqaured error is not that much if the overall samples tend to be 1000000000000. However, with no proper reference this metric cannot be used as a good representation of how impactful this mean squared error is. R2 however, is a good indicator to show how linear a model is.

## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>Explain YOUR PROCESS here:</b></font>

Where did you source your code?
I sourced my code primarily from the labs as well as from the sci-kit learn website which contained open source information about these libraries. Additionally, I had to look at Pandas notation to create the indexes for results. I utilized https://pandas.pydata.org/docs/user_guide/advanced.html to learn more about indexing within a Pandas' dataframe.

In what order did you complete the steps?
I completed the steps in the order they were presented. This is because it is the most logical way to process the data. First I imported the necessary libaries, then I intialized the model, afterwards I trained and split the data into train and test data sets. Afterwards we make the model fit the training dataset, afterwards we utilize our model to predict the values for the y values given the x testing dataset. We then compare the predicted y results to the actual y test data. Afterwards we utilize the mean squared error and R2 methods from sklearn to analyze the data and the efficacy of our model. Afterwards I simply had to organize this data in the results data frame from Pandas.

If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
I did not use generative AI for this section of the assignment since it built of the previous section. The main difference was the utilization of a different model. Here we used a linear model.

Did you have any challenges? If yes, what were they? If not, what helped you to be successful?
The main challenge I faced was indexing the information within result as desired. I utilized the pandas documentation online at the aforementioned link. 

# **Part 3: Observations/Interpretation (3 marks)**

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


<font color='Green'><b>ADD YOUR FINDINGS HERE</b></font>

A pattern that is reflected in the results found in the first section demonstrate the importance of proper feature selection and large data size which include lots of samples and lots of features within those samples. In the first Task when we had three different dataset collections and chose the first two columns in x to create our model and predict a value for y given a certain feature vector given for that model demonstrates that regardless of the amount of data some features are not great to build models on. This could have been due to the way the Logistical Regression model was created however, it could also be due to the irrelevance of the features. This could have been attributed to extreme noise within the datasets which caused these large model inaccuracies. This respective models training and validation accuracy were respectively, 0.616304 and 0.601087. This indicates that the model does not overfit the data since the training dataset did not have a very high accuracy. Thus it could likely be caused due to noise or irrelevant modeling of features which are not good predictors. This overall suggests that model has very high bias due to equally poor accuracies, meaning the model is heavily underperforming. Additionally, the first task demonstrated that feature selection is arguably more important than dataset sample size. If our model tries to build on irrelevant features, it does not matter the amount of data we have since it will still be predicting off poor indicators which don't factor in more important features. This is reflected once again with model 1 and 2 in the first task. X and y and X with the first two columns and y, had similar size of data samples however, the second model had a reduced feature vector( less features) and it performed significantly poorer. Their respective training and validation accuracies are as follows: 0.931957 and 0.942391, and 0.616304 and 0.601087. The third datasets model which had a smaller data size still performed better than the second model despite the large dataset difference. The third model's training and validation accuracies are as follows: 0.947826 and 0.905034. This could be attributed to more bias within this model since both the training and validation accuracy are high despite lower data samples. Ideally, more datasamples relates to better performance since this leads us to be closer to that of the absolute dataset. (The dataset which encompasses all scenarios)

In the second portion of the lab we created a linear regression model and then analyzed the model's mean squared error and R2. These two metrics are used to measure the performance of the model, specifically, R2 or the coefficent of determination, is utilized to view how linear the model's dataset really is. Thus it is a good measure to calculate if a linear model is adequate for the model. Often times though this metric could be disrupted by noise, bias, and other modeling which is more accurate. The R2 values we obtained for the testing and validation were respectively: 0.609071 and 0.636898. In my opinion this is very far from the ideal 1. Thus the data is not perfectly linear however, once again this could be attributed due to noise, in the data. It would be beneficial in the future to train the model on more data to help mitigate this data or alternatively, remove the outliers. Additionally, we calculated their respective Mean Squared Error, this measure finds the average error from the predicted and actual results generated from the model on its training dataset and then once again on a validation dataset. The respective values for the training and validation mean squared error were respectively, 110.345501 and 95.635335. Without a reference of the magnitude of the samples this number is not adequate enough. A percentage would need to be calculated to better interpolate the effects of this data. That being said though it does show that the model tended to perform better on the validation dataset than the training which is great because it shows that the model did not overfit. This could have been caused by samples with noise in the model. Perhaps in the future these datasets could be removed.

# **Part 4: Reflection (2 marks)**
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


<font color='Green'><b>ADD YOUR THOUGHTS HERE</b></font>
I really liked this assignment because it was an awesome application of the knowledge we learn which did not feel rushed as it does in labs. Additionally, I would like more guidance regarding what the questions would like us to discuss with regards to topics learned in class as well as an estimated desired depth in the answers we provide. It would be sick if assignments gave us bonus supplemental resources such as readings or assignments that we can do in our free time. Overall, I really enjoyed this assignment.

# **Part 5: Bonus Question (4 marks)**

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [22]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge, Lasso
from sklearn.metrics import r2_score, mean_squared_error

X_bonus, y_bonus = load_concrete()

X_train_bonus, X_test_bonus, y_train_bonus, y_test_bonus = train_test_split(X_bonus, y_bonus, test_size=0.2, random_state=42)

resultsRidge = pd.DataFrame(columns=['Alpha', 'Train_R2', 'Train_MSE', 'Val_R2', 'Val_MSE'])
resultsLasso = pd.DataFrame(columns=['Alpha', 'Train_R2', 'Train_MSE', 'Val_R2', 'Val_MSE'])

best_ridge_r2 = -np.inf
best_ridge_alpha = None

# Ridge regression 
for alpha in np.logspace(-3, 2, num=6):
    ridge = Ridge(alpha=alpha)
    ridge.fit(X_train_bonus, y_train_bonus)
    y_pred_train = ridge.predict(X_train_bonus)
    y_pred_val = ridge.predict(X_test_bonus)
    train_r2 = r2_score(y_train_bonus, y_pred_train)
    val_r2 = r2_score(y_test_bonus, y_pred_val)
    train_mse = mean_squared_error(y_train_bonus, y_pred_train)
    val_mse = mean_squared_error(y_test_bonus, y_pred_val)
    
    resultsRidge = resultsRidge.append({'Alpha': alpha, 'Train_R2': train_r2, 'Train_MSE': train_mse, 'Val_R2': val_r2, 'Val_MSE': val_mse}, ignore_index=True)

    if val_r2 > best_ridge_r2:
        best_ridge_r2 = val_r2
        best_ridge_alpha = alpha
print(resultsRidge)

print("Best R^2 score for Ridge regression:", best_ridge_r2)
print("Best alpha for Ridge regression:", best_ridge_alpha)

# Lasso regression
best_lasso_r2 = -np.inf
best_lasso_alpha = None

# Lasso regression
for alpha in np.logspace(-3, 2, num=6):
    lasso = Lasso(alpha=alpha)
    lasso.fit(X_train_bonus, y_train_bonus)
    y_pred_train = lasso.predict(X_train_bonus)
    y_pred_val = lasso.predict(X_test_bonus)
    train_r2 = r2_score(y_train_bonus, y_pred_train)
    val_r2 = r2_score(y_test_bonus, y_pred_val)
    train_mse = mean_squared_error(y_train_bonus, y_pred_train)
    val_mse = mean_squared_error(y_test_bonus, y_pred_val)
    
    resultsLasso = resultsLasso.append({'Alpha': alpha, 'Train_R2': train_r2, 'Train_MSE': train_mse, 'Val_R2': val_r2, 'Val_MSE': val_mse}, ignore_index=True)

    if val_r2 > best_lasso_r2:
        best_lasso_r2 = val_r2
        best_lasso_alpha = alpha

print(resultsLasso)
print("Best R^2 score for Lasso regression:", best_lasso_r2)
print("Best alpha for Lasso regression:", best_lasso_alpha)


     Alpha  Train_R2   Train_MSE    Val_R2    Val_MSE
0    0.001  0.610459  110.661771  0.627542  95.975484
1    0.010  0.610459  110.661771  0.627542  95.975484
2    0.100  0.610459  110.661771  0.627542  95.975480
3    1.000  0.610459  110.661771  0.627542  95.975441
4   10.000  0.610459  110.661772  0.627543  95.975055
5  100.000  0.610459  110.661862  0.627558  95.971315

C:\Users\de_ro\AppData\Local\Temp\ipykernel_3480\2089439915.py:28: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  resultsRidge = resultsRidge.append({'Alpha': alpha, 'Train_R2': train_r2, 'Train_MSE': train_mse, 'Val_R2': val_r2, 'Val_MSE': val_mse}, ignore_index=True)
C:\Users\de_ro\AppData\Local\Temp\ipykernel_3480\2089439915.py:28: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  resultsRidge = resultsRidge.append({'Alpha': alpha, 'Train_R2': train_r2, 'Train_MSE': train_mse, 'Val_R2': val_r2, 'Val_MSE': val_mse}, ignore_index=True)
C:\Users\de_ro\AppData\Local\Temp\ipykernel_3480\2089439915.py:28: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  resultsRidge = resultsRidge.append({'Alpha': alpha, 'Train_R2': train_r2, 'Train_M


Best R^2 score for Ridge regression: 0.6275577870808139
Best alpha for Ridge regression: 100.0
     Alpha  Train_R2   Train_MSE    Val_R2     Val_MSE
0    0.001  0.610459  110.661771  0.627542   95.975424
1    0.010  0.610459  110.661777  0.627544   95.974885
2    0.100  0.610457  110.662368  0.627563   95.969955
3    1.000  0.610250  110.721351  0.627577   95.966322
4   10.000  0.604579  112.332397  0.621375   97.564540
5  100.000  0.465346  151.885877  0.439762  144.362742
Best R^2 score for Lasso regression: 0.6275771642976664
Best alpha for Lasso regression: 1.0


C:\Users\de_ro\AppData\Local\Temp\ipykernel_3480\2089439915.py:53: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  resultsLasso = resultsLasso.append({'Alpha': alpha, 'Train_R2': train_r2, 'Train_MSE': train_mse, 'Val_R2': val_r2, 'Val_MSE': val_mse}, ignore_index=True)
C:\Users\de_ro\AppData\Local\Temp\ipykernel_3480\2089439915.py:53: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  resultsLasso = resultsLasso.append({'Alpha': alpha, 'Train_R2': train_r2, 'Train_MSE': train_mse, 'Val_R2': val_r2, 'Val_MSE': val_mse}, ignore_index=True)
C:\Users\de_ro\AppData\Local\Temp\ipykernel_3480\2089439915.py:53: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  resultsLasso = resultsLasso.append({'Alpha': alpha, 'Train_R2': train_r2, 'Train_M

<font color='Green'><b>ADD YOUR ANSWER HERE</b></font>
Which method and what value of alpha gave you the best R^2 score? 
Is this score "good enough"? Explain why or why not.

The method which gave me the best R^2 Score was the Ridge Model which yielded the best R^2 Score value for validation of approximately 0.6275577870808139 with an alpha of 100.0 with a training R^2 Score value of 0.610459. This score is still not good enough in my opinion since there is still a very large gap between that and an ideal model of 1.0. The ridge model applies a penalty determined by alpha to the least squares objective function which is then minimized to obtain the regression's coefficients. Perhaps it would be a good idea to try and obtain more data or potentially investigate the data to identify sources of noise and bias within the model created. This would help us better understand where the error or bias could be potentially be created.

https://www.andreaperlato.com/mlpost/deal-multicollinearity-with-ridge-regression/#:~:text=Ridge%20Regression%20performs%20a%20L2,the%20impact%20of%20correlated%20predictors.
